Source/WebCore/rendering/RenderLayer.cpp

```diff
@@ -1234,7 +1234,15 @@ void RenderLayer::recursiveUpdateLayerPositions(OptionSet<UpdateLayerPositionsFl
     }
 
     auto repaintIfNecessary = [&](bool checkForRepaint) {
-        if (isVisibilityHiddenOrOpacityZero() || !isSelfPaintingLayer()) {
+        if (isVisibilityHiddenOrOpacityZero()) {
+            LAYER_POSITIONS_ASSERT_IMPLIES(mode == Verify, !repaintRects());
+            clearRepaintRects();
+            return;
+        }
+
+        if (!isSelfPaintingLayer()) {
+            if (checkForRepaint && shouldRepaintAfterLayout())
+                repaintIncludingNonCompositingDescendants(renderer().containerForRepaint().renderer.get());
```

Comment:
Could we have found this repaint container earlier and told it to repaint, rather than adding `NeedsFullRepaintEvenIfNotSelfPainting`?

Reply:
It's not the container, it's the rects to use with it.

When something stops compositing (during a style flush), we don't really know what rects to mark for repaint in the nearest compositing ancestor, since layout is probably stale.

The current code sets the `NeedsFullRepaint` flag so that we mark a repaint after the next layout, but this gets skipped if the RenderLayer isn't self painting (which must be a new change at the same time as it stopped compositing).

The core change here is adding a new flag for this case, so that the post-layout render layer update issues the repaint, even for things that are no longer self-painting.